### PR TITLE
Automatically restart analysis server

### PR DESCRIPTION
**Pull request by lukechurch**
This automatically restarts the analysis server if a request times out and improves logging

TBR @devoncarew

### CODE REVIEW

File: lib/src/common_server.dart [hunk 1]
```diff
@@ -67,28 +67,32 @@ class CommonServer {
   AnalysisServerWrapper analysisServer;
   AnalysisServerWrapper analysisServerStrong;
 
-  CommonServer(String sdkPath, this.container, this.cache,
+  String sdkPath;
```
Comment (devoncarew) [inline lib/src/common_server.dart]: final?

File: lib/src/common_server.dart [hunk 2]
```diff
@@ -67,28 +67,32 @@ class CommonServer {
...
+
+  CommonServer(String this.sdkPath, this.container, this.cache,
```
Comment (devoncarew) [inline lib/src/common_server.dart]: you can remove the `String` annotation

File: lib/src/common_server.dart [hunk 3]
```diff
@@ -101,6 +105,17 @@ class CommonServer {
     await analysisServerStrong.warmup(useHtml: useHtml);
   }
 
+  Future restart() async {
+    _logger.warning('Restarting CommonServer');
+    await shutdown();
+    _logger.info('Analysis Servers shutdown');
```
Comment (devoncarew) [inline lib/src/common_server.dart]: sp: Server(s)

File: lib/src/common_server.dart [hunk 4]
```diff
@@ -458,14 +475,20 @@ class CommonServer {
...
+          sources,
+          new Location()
+            ..sourceName = sourceName
+            ..offset = offset);
+      _logger.info(
+          'PERF: Computed completions in ${watch.elapsedMilliseconds}ms.');
+      return response;
+    } catch (e, st) {
+      _logger.severe('Error during _complete', e, st);
+      await restart();
```
Comment (devoncarew) [inline lib/src/common_server.dart]: 👍

File: lib/src/common_server.dart (hunk 1 above)
Comment (lukechurch) [inline lib/src/common_server.dart]: thanks, will fix

File: lib/src/common_server.dart (hunk 2 above)
Comment (lukechurch) [inline lib/src/common_server.dart]: thanks, will fix

File: lib/src/common_server.dart (hunk 3 above)
Comment (lukechurch) [inline lib/src/common_server.dart]: Bizarrely there are actually two of them (strong mode and non-strong mode). I keep forgetting this as well.

**Pull request merged**